File: src/claw_vault/claude_code/runtime_action_client.py
"""Claude Code Runtime Action Guard client."""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any

import httpx

from claw_vault.guard.runtime_action import RuntimeDecision, RuntimeRiskLevel


@dataclass(frozen=True)
class RuntimeActionClientDecision:
    """Runtime Action Guard decision returned to a Claude Code caller."""

    decision: str
    risk_level: str
    reasons: list[str]
    categories: list[str]
    action_type: str
    target_summary: str
    redacted_summary: str
    should_block: bool
    block_reason: str | None = None
    audit_recorded: bool = False


@dataclass(frozen=True)
class ClaudeCodeRuntimeActionClient:
    """Client for the local ClawVault Runtime Action Guard API."""

    clawvault_url: str = "http://127.0.0.1:8766"
    request_timeout_seconds: float = 2.0
    source_agent: str = "claude-code"
    _client: httpx.AsyncClient | None = field(default=None, repr=False, compare=False)

    async def evaluate(
        self,
        *,
        tool_name: str,
        params: dict[str, Any],
        agent_id: str | None = None,
        session_id: str | None = None,
    ) -> RuntimeActionClientDecision:
        payload = {
            "source_agent": self.source_agent,
            "tool_name": tool_name,
            "params": params,
            "agent_id": agent_id,
            "session_id": session_id,
        }
        try:
            data = await self._post(payload)
            return _parse_decision(data, tool_name)
        except Exception:
            return _fail_closed(tool_name, "Runtime Action Guard unavailable")

    async def _post(self, payload: dict[str, Any]) -> dict[str, Any]:
        if self._client is not None:
            response = await self._client.post("/api/runtime-action", json=payload)
            response.raise_for_status()
            data = response.json()
            if isinstance(data, dict):
                return data
            raise ValueError("malformed runtime action decision")

        timeout = httpx.Timeout(self.request_timeout_seconds)
        async with httpx.AsyncClient(
            base_url=self.clawvault_url.rstrip("/"),
            timeout=timeout,
        ) as client:
            response = await client.post("/api/runtime-action", json=payload)
            response.raise_for_status()
            data = response.json()
            if isinstance(data, dict):
                return data
            raise ValueError("malformed runtime action decision")


def _parse_decision(data: dict[str, Any], tool_name: str) -> RuntimeActionClientDecision:
    if (
        not _is_decision(data.get("decision"))
        or not _is_risk_level(data.get("risk_level"))
        or not isinstance(data.get("reasons"), list)
        or not isinstance(data.get("categories"), list)
        or not isinstance(data.get("action_type"), str)
        or not isinstance(data.get("target_summary"), str)
        or not isinstance(data.get("redacted_summary"), str)
        or not isinstance(data.get("should_block"), bool)
        or not isinstance(data.get("audit_recorded"), bool)
    ):
        raise ValueError("malformed runtime action decision")

    if data["decision"] != RuntimeDecision.ALLOW.value and data["should_block"] is not True:
        raise ValueError("unsafe runtime action decision")

    block_reason = data.get("block_reason")
    return RuntimeActionClientDecision(
        decision=data["decision"],
        risk_level=data["risk_level"],
        reasons=[item for item in data["reasons"] if isinstance(item, str)],
        categories=[item for item in data["categories"] if isinstance(item, str)],
        action_type=data["action_type"],
        target_summary=data["target_summary"],
        redacted_summary=data["redacted_summary"],
        should_block=data["should_block"],
        block_reason=block_reason if isinstance(block_reason, str) else None,
        audit_recorded=data["audit_recorded"],
    )


def _fail_closed(tool_name: str, reason: str) -> RuntimeActionClientDecision:
    return RuntimeActionClientDecision(
        decision=RuntimeDecision.BLOCK.value,
        risk_level=RuntimeRiskLevel.CRITICAL.value,
        reasons=[reason],
        categories=["runtime_action_guard_unavailable"],
        action_type=_fallback_action_type(tool_name),
        target_summary="",
        redacted_summary="",
        should_block=True,
        block_reason=f"ClawVault Runtime Action Guard unavailable for {tool_name}",
        audit_recorded=False,
    )


def _fallback_action_type(tool_name: str) -> str:
    lowered = tool_name.lower()
    if lowered == "bash":
        return "shell.execute"
    if lowered == "read":
        return "file.read"
    if lowered in {"write", "edit"}:
        return "file.write"
    if lowered in {"webfetch", "websearch"}:
        return "network.request"
    return "tool.call"


def _is_decision(value: object) -> bool:
    return value in {item.value for item in RuntimeDecision}


def _is_risk_level(value: object) -> bool:
    return value in {item.value for item in RuntimeRiskLevel}

File: openclaw-file-guard/src/runtime-action-client.ts
import axios, { type AxiosInstance } from "axios";
import type { AgentContext, Logger, PluginRuntimeConfig, ToolCallEvent } from "./types.js";

export interface RuntimeActionDecision {
  decision: "allow" | "block" | "ask-user";
  risk_level: "low" | "medium" | "high" | "critical";
  reasons: string[];
  categories: string[];
  action_type: string;
  target_summary: string;
  redacted_summary: string;
  should_block: boolean;
  block_reason?: string | null;
  audit_recorded: boolean;
}

export interface RuntimeActionFailure {
  decision: "block";
  risk_level: "critical";
  reasons: string[];
  categories: string[];
  action_type: string;
  target_summary: string;
  redacted_summary: string;
  should_block: true;
  block_reason: string;
  audit_recorded: false;
}

export type RuntimeActionResult = RuntimeActionDecision | RuntimeActionFailure;

export class RuntimeActionClient {
  private readonly http: AxiosInstance;

  constructor(
    private readonly config: PluginRuntimeConfig,
    private readonly logger: Logger,
  ) {
    this.http = axios.create({
      baseURL: config.clawvaultUrl.replace(/\/+$/, ""),
      timeout: config.requestTimeoutMs,
    });
  }

  supports(event: ToolCallEvent): boolean {
    const toolName = event.toolName.toLowerCase();
    return toolName === "bash" || toolName === "read" || toolName === "write";
  }

  async evaluate(
    event: ToolCallEvent,
    ctx: AgentContext,
  ): Promise<RuntimeActionResult> {
    try {
      const res = await this.http.post("/api/openclaw/runtime-action", {
        tool_name: event.toolName,
        params: event.params ?? {},
        agent_id: ctx?.agentId,
        session_id: ctx?.sessionId ?? ctx?.sessionKey,
      });
      return parseDecision(res.data, event.toolName);
    } catch (err) {
      const msg = err instanceof Error ? err.message : String(err);
      this.logger.warn(`runtime action guard unavailable: ${msg}`);
      return failClosed(event.toolName, "Runtime Action Guard unavailable");
    }
  }
}

function parseDecision(data: unknown, toolName: string): RuntimeActionDecision {
  if (!data || typeof data !== "object") {
    throw new Error("malformed runtime action decision");
  }
  const value = data as Partial<RuntimeActionDecision>;
  if (
    !isDecision(value.decision) ||
    !isRiskLevel(value.risk_level) ||
    !Array.isArray(value.reasons) ||
    !Array.isArray(value.categories) ||
    typeof value.action_type !== "string" ||
    typeof value.target_summary !== "string" ||
    typeof value.redacted_summary !== "string" ||
    typeof value.should_block !== "boolean" ||
    typeof value.audit_recorded !== "boolean"
  ) {
    throw new Error("malformed runtime action decision");
  }
  if (value.decision !== "allow" && value.should_block !== true) {
    throw new Error("unsafe runtime action decision");
  }
  return {
    decision: value.decision,
    risk_level: value.risk_level,
    reasons: value.reasons.filter((item): item is string => typeof item === "string"),
    categories: value.categories.filter((item): item is string => typeof item === "string"),
    action_type: value.action_type,
    target_summary: value.target_summary,
    redacted_summary: value.redacted_summary,
    should_block: value.should_block,
    block_reason: value.block_reason,
    audit_recorded: value.audit_recorded,
  };
}

function failClosed(toolName: string, reason: string): RuntimeActionFailure {
  return {
    decision: "block",
    risk_level: "critical",
    reasons: [reason],
    categories: ["runtime_action_guard_unavailable"],
    action_type: toolName.toLowerCase() === "bash" ? "shell.execute" : `tool.${toolName.toLowerCase()}`,
    target_summary: "",
    redacted_summary: "",
    should_block: true,
    block_reason: `ClawVault Runtime Action Guard unavailable for ${toolName}`,
    audit_recorded: false,
  };
}

function isDecision(value: unknown): value is RuntimeActionDecision["decision"] {
  return value === "allow" || value === "block" || value === "ask-user";
}

function isRiskLevel(value: unknown): value is RuntimeActionDecision["risk_level"] {
  return value === "low" || value === "medium" || value === "high" || value === "critical";
}
